Implement VirtIO GPU and input devices

shengwen-tw opened the pull request:
## Summary

This pull request introduces initial support for `virtio-gpu` and `virtio-input` devices in `semu`, conforming to the [VirtIO specification 1.3](https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.html).

### VirtIO-GPU

* Based on `VirtIO MMIO transport layer` (i.e., Linux platform device)
* Implements basic 2D rendering support.
* Provides two rendering backends:
    * **Software-based** backend (supports 2D rendering only, ready to be integrated into [rv32emu](https://github.com/sysprog21/rv32emu)).
    * **virglrenderer** backend (to enable hardware-accelerated 3D rendering in a future patch).
* Compatible with:
    * [DirectFB2](https://github.com/directfb2/DirectFB2)
    * [X11 Window System](https://www.x.org/wiki/)
* Successfully tested with:
    * [kmscube](https://gitlab.freedesktop.org/mesa/kmscube)
    * [glmark2](https://github.com/glmark2/glmark2)
    * [glxgears](https://linuxreviews.org/Glxgears#google_vignette)

### VirtIO-Input

* Implements basic keyboard emulation.
* Implements basic mouse emulation.

## Prerequisites

> Note: this list is not yet exhaustive

```shell
$ sudo apt install libsdl2-dev libsdl2-2.0-0 libsdl2-image-dev libsdl2-image-2.0-0
```

## Build kernel image and test rootfs image

```shell
$ make rootfs.cpio # Download prebuilt minimal rootfs for initrd
$ ./scripts/build-image.sh --all --x11 --external-root --extra-packages
```

Use `--x11` for full setup; omit it for a minimal setup.

## Run `X11`

1. Launch `semu`:

```shell
$ make check
```

2. Run `startx`

```
$ startx
```

3. After `X11` shows up, launch the benchmark programs:

```
$ glmark2
$ glxgears
```

## Run `DirectFB2`

1. Launch `semu`:

```shell
$ make check
```

2. Kill `X11` (Skip if `--x11` is not used)

```
$ ps | grep 'Xorg'
   <PID> root     /usr/bin/Xorg :0.0 vt01 -s 0 -noreset -allowMouseOpenFail
$ kill <PID>
```

3. Set up environment variables via `run.sh`:

```shell
$ source ./run.sh
```

4. Run `DirectFB2` test programs:

```
df_andi        df_fire df_neo     df_window
df_texture     df_video           df_knuckles
df_input       df_neo             df_window
df_matrix      df_dok             df_fire
df_palette     df_cpuload         df_spacedream
df_layers      df_pss             df_input
```

## Run `kmscube`

1. Launch `semu`:

```shell
$ make check
```

2. Kill `X11` (Skip if `--x11` is not used)

```
$ ps | grep 'Xorg'
   <PID> root     /usr/bin/Xorg :0.0 vt01 -s 0 -noreset -allowMouseOpenFail
$ kill <PID>
```

3. Run `kmscube`

```
$ kmscube
```

## Run `modetest`

The simplest method to test VirtIO-GPU.

1. Enable `BR2_PACKAGE_LIBDRM_INSTALL_TESTS` in the Buildroot then rebuild the image.

2. Launch `semu`:

```shell
$ make check
```

3. Run `modetest` program

```shell
$ modetest -M virtio_gpu -s 34:1024x768
```

A test image should then show up in the display window.

jserv commented:
How about using [virglrenderer](https://gitlab.freedesktop.org/virgl/virglrenderer)?
See https://mail.gnu.org/archive/html/qemu-devel/2021-02/msg04235.html

shengwen-tw commented:
Yes, I do plan to support VirGL for acceleration.

As far as I know, the offloading of rendering operations to the host GPU is called 3D mode by VirtIO's spec, whereas this pull request only implemented 2D mode due to the overall complexity.

Before implementing 3D rendering mode with VirGL, I think we need to configure the Buildroot to have a minimal desktop environment or set the Mesa benchmark tools without a desktop manager (I'm not sure if this is feasible).

In conclusion, I'll recommend first merging this pull request as a checkpoint for virtio-gpu and then adding support for VirGL in a later pull request.

jserv reviewed (changes requested):
Rebase and resolve the conflicts.

jserv commented on README.md:
````diff
@@ -37,6 +40,28 @@ For macOS, use the following command:
 $ brew install e2fsprogs
 ```
 
+### SDL2 Installation
````
It is not necessary to build SDL2 and pixman. Let's make use of prebuilt packages.

shengwen-tw commented:
Reference: [How to run headless unit tests for GUIs on GitHub actions](https://arbitrary-but-fixed.net/2022/01/21/headless-gui-github-actions.html)

jserv commented on Makefile:
```diff
@@ -36,6 +39,37 @@ ifeq ($(call has, VIRTIONET), 1)
+endif
+
+# Pixman
+ENABLE_PIXMAN ?= 1
```
Pixman might be absent for some platforms such as macOS and Windows. Can you get rid of it?

shengwen-tw replied:
It seems like [Homebrew](https://formulae.brew.sh/formula/pixman) does exist a pre-built package of Pixman. (For Windows, the problem seems a bit tricky)

If we decide to handle the low-level pixel manipulation by ourselves, we may need to support various image formats as shown in [here](https://github.com/shengwen-tw/semu/blob/master/virtio-gpu.c#L292).
Please let me know about your opinion.

jserv replied:
Should the requirement be limited to performing color space conversion, we can offer a streamlined implementation specifically for this purpose.

shengwen-tw replied:
It turns out that it requires not much modification to remove the dependency of the Pixman.
We only need to allocate the image buffer properly and manage the stride size and image depth (i.e., Bits per Pixel).

jserv commented:
Rather than relying on Xvfb as a heavy-duty solution for validating Linux DRM/framebuffer, it might be more efficient to use lighter tools like [drm-framebuffer](https://github.com/embear-engineering/drm-framebuffer). This tool, designed for testing drm devices, takes data from stdin and displays it, offering a simpler and more straightforward approach.

shengwen-tw commented:
> Rather than relying on Xvfb as a heavy-duty solution for validating Linux DRM/framebuffer, it might be more efficient to use lighter tools like [drm-framebuffer](https://github.com/embear-engineering/drm-framebuffer). This tool, designed for testing drm devices, takes data from stdin and displays it, offering a simpler and more straightforward approach.

Actually, the idea of using Xvfb here is to make sure the SDL window can be launched in the CI environment by GitHub Action.

jserv commented on Makefile:
```diff
@@ -36,6 +39,28 @@ ifeq ($(call has, VIRTIONET), 1)
+endif
+
+# SDL2
+ENABLE_SDL ?= 1
```
The virtio-gpu support depends on SDL, so that you have to check the rationale.

shengwen-tw replied:
Resolved.

jserv commented on window.h:
```diff
@@ -0,0 +1,15 @@
+
+#include <SDL.h>
+
+void display_window_init(void);
```
Change the prefix for functions specific to window management to "window_" from "display_window."

shengwen-tw replied:
Resolved.

jserv commented on Makefile:
```diff
@@ -1,13 +1,16 @@
 
 CC ?= gcc
-CFLAGS := -O2 -g -Wall -Wextra
+CFLAGS := -O2 -g -Wall -Wextra -Wno-unused-function
```
Don't change this flag. We should always check unused functions.

shengwen-tw replied:
Resolved.

jserv commented on Makefile:
```diff
@@ -36,6 +39,28 @@ ifeq ($(call has, VIRTIONET), 1)
+$(warning No sdl2-config in $$PATH. Check SDL2 installation in advance)
+override ENABLE_SDL := 0
+endif
+CFLAGS += `sdl2-config --cflags`
```
Use `$(shell sdl2-config --cflags)` instead.

shengwen-tw replied:
Resolved.

jserv commented on configs/linux.config:
```diff
@@ -1,3 +1,7 @@
+#
+# Automatically generated file; DO NOT EDIT.
+# Linux/riscv 6.1.55 Kernel Configuration
+#
```
Remove these lines.

shengwen-tw replied:
Resolved.

jserv commented on configs/linux.config:
```diff
@@ -842,7 +867,45 @@ CONFIG_DEVMEM=y
 # I2C support
 #
-# CONFIG_I2C is not set
+CONFIG_I2C=m
```
See if i2c configuration is required.

shengwen-tw replied:
The I2C could not be disabled as it is required by the DRM module even if virtio-gpu doesn't really on it.

jserv commented:
Merely executing `Xvfb :1 -screen 0 1600x1200x24` is inadequate because, although it confirms the initialization of SDL with libX11, it does not guarantee the accuracy of the window content rendered in Xvfb.

shengwen-tw commented on window.c:
```diff
@@ -0,0 +1,137 @@
+    SDL_UnlockMutex(displays[id].img_mtx);
+}
+
+void display_window_render(uint32_t id,
```
Check [[1]](https://github.com/shengwen-tw/semu/blob/master/virtio-gpu.c#L40) and [[2]](https://github.com/shengwen-tw/semu/blob/master/window.c#L131).

shengwen-tw commented:
> Merely executing `Xvfb :1 -screen 0 1600x1200x24` is inadequate because, although it confirms the initialization of SDL with libX11, it does not guarantee the accuracy of the window content rendered in Xvfb.

The latest modification allows the virtio-gpu to be run without SDL, so the current CI test can pass by executing `make ENABLE_SDL=0`.

jserv commented on virtio-gpu.c:
```diff
@@ -0,0 +1,1084 @@
+        return -1;
+
+    /* Release the resource */
+    if (res_2d->image)
```
Remove this `if` statement.

shengwen-tw replied:
Resolved.

jserv commented on virtio-gpu.c:
```diff
@@ -0,0 +1,1084 @@
+            list_entry(curr, struct vgpu_resource_2d, list);
+
+        list_del(&res_2d->list);
+        if (res_2d->image)
```
Quote from Linux Programmer's Manual
> If ptr is NULL, no operation is performed.

Remove unnecessary `if` statement before the invocation of `free`.

shengwen-tw replied:
Resolved.

jserv commented on window.c:
```diff
@@ -0,0 +1,169 @@
+    display_cnt++;
+}
+
+int window_thread(void *data)
```
`window_thread` is not a public function. Make it private by declaring `static`.

shengwen-tw replied:
Resolved.

jserv commented on window.h:
```diff
@@ -0,0 +1,11 @@
+void window_init(void);
+void window_add(uint32_t width, uint32_t height);
+void window_render(void *resource);
+void display_resource_lock(uint32_t id);
```
Rename to `window_lock` and `window_unlock`.

jserv commented on configs/buildroot.config:
```diff
@@ -1,14 +1,62 @@
+BR2_HAVE_DOT_CONFIG=y
```
Put only the necessary configuration items in this file. Minimize the necessary changes.

jserv replied:
I am concerned about the increasing size of both the Linux kernel and the root file system. The suggested updates allow the emulator to be tailored based on selected features like virtio-gpu. However, from the viewpoint of buildroot, it involves a single build process that decides on the package compilation without recognizing the configurable features.

shengwen-tw replied:
I received the following error after attempting to modify the intrinsic parameter manually:

```
$ make
Makefile:943: *** Please configure Buildroot first (e.g. "make menuconfig").  Stop.
make: *** [Makefile:82: _all] Error 2
```

It seems like the Buildroot doesn't want to allow the user to modify the `.config` file without `make menuconfig`.

jserv replied:
Of course, you shall keep `BR2_HAVE_DOT_CONFIG`. Meanwhile, you can shrink the lines of configuration file.

shengwen-tw replied:
The following three lines are the essential settings for enabling DRM in Builtroot:

```
BR2_PACKAGE_LIBDRM=y
BR2_PACKAGE_LIBDRM_HAS_ATOMIC=y
BR2_PACKAGE_LIBDRM_INSTALL_TESTS=y
```

However, after executing `make olddefconfig`, the Buildroot will still append several lines in `.config` to keep the configuration file up to date of the latest version of Buildroot.

jserv replied:
Strip `buildroot.config` and utilize `oldconfig` to merge the proposed changes.

jserv commented on scripts/run.sh:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/bash
```
All files in the scripts directory are implicitly designed to run on the host system. Conversely, if a file is specifically intended for the RV32 target, it should not be placed in this directory.

shengwen-tw replied:
`run.sh` is now relocated to the newly created directory `guest/` .

jserv commented on scripts/run.sh:
```diff
@@ -0,0 +1,10 @@
+mkdir -p /lib/modules/$(uname -r)
+cp ./modules/*.ko /lib/modules/$(uname -r)
+
+depmod -a
```
It is not necessary to run `depmon` inside RV32 target. Instead, check https://unix.stackexchange.com/questions/401138

shengwen-tw replied:
Resolved by invoking `make install_modules` to build the `modules.dep` file. (d9472b0553eab900cfcb753ffc7910c83c1c4735)

jserv commented on window.c:
```diff
@@ -0,0 +1,169 @@
+#include <pthread.h>
```
For the sake of portability, don't include `<pthread.h>` or `<semaphore.h`>. Instead, place `#include <SDL_thread.h>` as a generic solution.

shengwen-tw replied:
Resolved.

jserv commented on window.c:
```diff
@@ -0,0 +1,169 @@
+    struct display_info *display = (struct display_info *) data;
+
+    /* Create SDL window */
+    display->window = SDL_CreateWindow("Semu", SDL_WINDOWPOS_UNDEFINED,
```
Use lowercase `semu`.

shengwen-tw replied:
Resolved.

jserv commented on window.c:
```diff
@@ -0,0 +1,169 @@
+
+void window_init(void)
+{
+    char thread_name[100] = {0};
```
It is not worthy to specify lengthy string.

shengwen-tw replied:
Resolved by shrinking the size from 100 to 20. Check [window.c#L98](https://github.com/shengwen-tw/semu/blob/master/window.c#L98).

jserv commented on window.c:
```diff
@@ -0,0 +1,169 @@
+        break;
+    }
+
+    displays[id].width = display->width;
```
Can you utilize `memcpy` for such initialization? Be aware of structure packing.

shengwen-tw replied:
Check 965aad4fe883d79c071a2457b7f6a99ca7cb2e46.

jserv commented on window.h:
```diff
@@ -0,0 +1,11 @@
+void window_init(void);
+void window_add(uint32_t width, uint32_t height);
+void window_render(void *resource);
+void window_resource_lock(uint32_t id);
```
Simply name it `window_lock` and `window_unlock` since their names already imply the shared resource management.

shengwen-tw replied:
Resolved.

jserv commented on Makefile:
```diff
@@ -36,6 +39,32 @@ ifeq ($(call has, VIRTIONET), 1)
+    override ENABLE_SDL := 0
+endif
+
+ifeq ($(ENABLE_VIRTIOGPU),1)
```
If SDL is absent, there should be no support for `ENABLE_VIRTIOGPU`.

shengwen-tw replied:
Resolved in 7985adc17c9a3a4e01f4a470dfba0851f3d2ed14.

jserv commented on window.c:
```diff
@@ -0,0 +1,171 @@
+    }
+}
+
+void window_render(void *_resource)
```
Avoid specifying the type of parameters to `void *` if the function is not designed for generic usage.

shengwen-tw replied:
Resolved in 12a76086abbdd9bbc10a0d3377b857cb54212aa8.

jserv commented:
I dislike the inclusion to target the X Window System for 2D rendering verification in minimalist system emulators like `semu`. We could instead leverage existing packages that render graphical content directly to the Linux framebuffer, supported by the proposed virtio-gpu path. On the other hand, the installation and footprint of the X Window System are generally larger, resulting in limited functionality with minimal installations.

shengwen-tw commented:
> I dislike the inclusion to target the X Window System for 2D rendering verification in minimalist system emulators like `semu`. We could instead leverage existing packages that render graphical content directly to the Linux framebuffer, supported by the proposed virtio-gpu path. On the other hand, the installation and footprint of the X Window System are generally larger, resulting in limited functionality with minimal installations.

How about using DirectFB with Mesa? Here is a [FOSDEM presentation](https://archive.fosdem.org/2022/schedule/event/directfb/attachments/slides/5077/export/events/attachments/directfb/slides/5077/fosdem_2022_directfb2.pdf) I found.
One of the reasons I included X Window System is because later we can test 3D rendering mode (i.e., VirGL) with it.

jserv commented:
> How about using DirectFB with Mesa? Here is a [FOSDEM presentation](https://archive.fosdem.org/2022/schedule/event/directfb/attachments/slides/5077/export/events/attachments/directfb/slides/5077/fosdem_2022_directfb2.pdf) I found. One of the reasons I included X Window System is because later we can test 3D rendering mode (i.e., VirGL) with it.

[DirectFB2](https://directfb2.github.io/) (the active DirectFB fork) is really amazing. If buildroot is capable of building DirectFB2 packages from scratch, I would like to integrate it as the major showcase.

jserv reviewed (changes requested):
Rebase the latest `master` branch.

jserv commented on configs/linux.config:
```diff
@@ -905,18 +1016,128 @@ CONFIG_MFD_SYSCON=y
+CONFIG_FB_NOTIFY=y
+CONFIG_FB=y
+# CONFIG_FIRMWARE_EDID is not set
+CONFIG_FB_CFB_FILLRECT=m
```
Drop framebuffer modules. Enable the minimal functionality for Linux framebuffer.

shengwen-tw replied:
I'm in an intense work period of my job (until ~July), so I will adjust the pull request after that.
By the way, `semu` can now run `DirectFB2` with default examples now, I will provide detailed information later.

shengwen-tw committed:
Build DirectFB2 as extra package

shengwen-tw committed:
Implement virtio-gpu device

shengwen-tw committed:
Disable SDL on GitHub Actions

jserv commented on scripts/build-image.sh:
```diff
@@ -38,16 +38,50 @@ function do_buildroot
+    export PATH=$PATH:$PWD/buildroot/output/host/bin
+    export BUILDROOT_OUT=$PWD/buildroot/output/
+    pushd DirectFB2
+    wget https://gist.githubusercontent.com/shengwen-tw/e5dffd8aefd7d2019cfbb4b7f1048ef3/raw/b079443aefb965742a6c6227fe4d0cf831d87cb1/riscv-directfb2.patch
```
Don't do that. You can create a directory named `patches` and place the corresponding patches against the packages you want to patch.

jserv replied:
Eventually, you should promote the DirectFB2 inclusion for the buildroot project. Get it adopted by upstream.

shengwen-tw replied:
> Don't do that. You can create a directory named `patches` and place the corresponding patches against the packages you want to patch.

Resolved.

shengwen-tw replied:
> Eventually, you should promote the DirectFB2 inclusion for the buildroot project. Get it adopted by upstream.

I'll tag the pull request link of DirectFB2 inclusion for Buildroot at #32 once it is ready.

jserv replied:
> I'll tag the pull request link of DirectFB2 inclusion for Buildroot at #32 once it is ready.

Good. In this repository, we aim to build packages minimally by using established tools like buildroot. Therefore, prioritizing upstream contributions is a fundamental principle for this purpose.

jserv commented on minimal.dts:
```diff
@@ -12,7 +12,8 @@
         bootargs = "earlycon console=ttyS0";
         stdout-path = "serial0";
-        linux,initrd-start = <0x1f700000>; /* @403 MiB (503 * 1024 * 1024)     */
+        /* XXX: 105MiB */
```
Refine the comment.

shengwen-tw replied:
Resolved.

jserv commented on guest/run.sh:
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/bash
```
Rename the directory `guest` to `target` as [kvm-host](https://github.com/sysprog21/kvm-host) does.

shengwen-tw replied:
Resolved.

shengwen-tw commented:
Git-pushed for progress saving.
`linux.config` and `buildroot.config` are still subjected to change (i.e., disable unused functionalities).
The proper size for `initrd` space will be calculated after size reduction.

shengwen-tw commented:
Discard changes for `buildroot.config` as no additional feature is needed for enabling DirectFB2 with virtio-gpu.

shengwen-tw commented:
Discard patch to the `DirectFB2` as the modification is accepted by the maintainer.

jserv commented:
We can enable SDL support through configuration changes, but the BusyBox-specific settings generate much larger filesystem images, unsuitable for minimal usage. Can you suggest an effective way to switch or customize these settings?

shengwen-tw commented:
> We can enable SDL support through configuration changes, but the BusyBox-specific settings generate much larger filesystem images, unsuitable for minimal usage. Can you suggest an effective way to switch or customize these settings?

I think we now have no modification to `BusyBox`'s setting now as they were dropped in an earlier version.
This pull request currently only modifies `linux.config` with the following new features enabled:
1. virtio-gpu driver
2. fbdev (Linux Framebuffer) subsystem
3. DRM subsystem (Required by virtio-gpu)
4. HDMI driver (Force selected by DRM)
5. I2C driver (Force selected by DRM)
6. DMA shared buffer feature (Force selected by DRM)

According to the Kconfig files from the Linux kernel:

```diff
config DRM_VIRTIO_GPU
        tristate "Virtio GPU driver"
>        depends on DRM && VIRTIO_MENU && MMU
        select VIRTIO
        select DRM_KMS_HELPER
        select DRM_GEM_SHMEM_HELPER
        select VIRTIO_DMA_SHARED_BUFFER
        help
           This is the virtual GPU driver for virtio.  It can be used with
           QEMU based VMMs (like KVM or Xen).

           If unsure say M.
```

```diff
menuconfig DRM
        tristate "Direct Rendering Manager (XFree86 4.1.0 and higher DRI support)"
        depends on (AGP || AGP=n) && !EMULATED_CMPXCHG && HAS_DMA
        select DRM_NOMODESET
        select DRM_PANEL_ORIENTATION_QUIRKS
>        select HDMI
>        select FB_CMDLINE
>        select I2C
        select I2C_ALGOBIT
>        select DMA_SHARED_BUFFER
        select SYNC_FILE
...
```

Also, since `SDL2` is compiled with `semu` as a part of the host, I think it is not related to `.config` file setting.

shengwen-tw commented:
A `Buildroot` patch that added support for `DirectFB2` is submitted via mailing list.
For the code modification, please refer to https://github.com/shengwen-tw/buildroot/commit/aad03b5b786028e61cf602677adfde2324a5939c.

jserv commented:
Can you provide a video recording demonstrating virtio-gpu support in RISC-V system emulation, showcasing graphics rendering and input interactions?

jserv commented on configs/linux.config:
```diff
@@ -748,30 +802,180 @@ CONFIG_INPUT=y
+# CONFIG_MOUSE_ELAN_I2C is not set
+# CONFIG_MOUSE_VSXXXAA is not set
+# CONFIG_MOUSE_SYNAPTICS_I2C is not set
+CONFIG_INPUT_JOYSTICK=y
```
Remove unused joystick.

jserv commented on configs/linux.config:
```diff
@@ -987,7 +1426,7 @@ CONFIG_HID_GENERIC=y
 # CONFIG_HID_MICROSOFT is not set
 # CONFIG_HID_MONTEREY is not set
-# CONFIG_HID_MULTITOUCH is not set
+CONFIG_HID_MULTITOUCH=y
```
Drop this.

jserv commented on virtio-input.c:
```diff
@@ -0,0 +1,642 @@
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <linux/input-event-codes.h>
```
Can you avoid the inclusion of Linux specific headers?

jserv replied:
That is, semu should be as portable as possible, capable of running on the environments based on Linux, macOS, and even Windows. If virtio-input is highly coupled with Linux host, there should be explicitly checks.

jserv commented on window.c:
```diff
@@ -0,0 +1,186 @@
+#include <linux/input-event-codes.h>
```
Ditto. Avoid the inclusion of Linux specific headers.

jserv commented on bitops.h:
```diff
@@ -0,0 +1,67 @@
+#ifndef __BITOPS_H__
```
Only keep the necessary functions and move them to `common.h`.

jserv commented on configs/linux.config:
```diff
@@ -224,7 +233,10 @@ CONFIG_ARCH_RV32I=y
 CONFIG_CMODEL_MEDANY=y
 CONFIG_MODULE_SECTIONS=y
 CONFIG_SMP=y
+CONFIG_NR_CPUS=32
```
Avoid changing SMP specific configurations.

jserv commented on input-event-codes.h:
```diff
@@ -0,0 +1,992 @@
+/* SPDX-License-Identifier: GPL-2.0-only WITH Linux-syscall-note */
```
Don't check in GPL-2.0-only code! Instead, extract the identifiers you need.

jserv commented on window.h:
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#if SEMU_HAS(VIRTIOGPU)
+#include <SDL.h>
```
This header has nothing related to SDL2.

jserv commented on Makefile:
```diff
@@ -60,6 +60,14 @@ ifneq ($(UNAME_S),Linux)
+# VirGL
+ENABLE_VIRGL ?= 0
+ifneq (ENABLE_VIRTIOGPU,0)
+    CFLAGS += `pkg-config virglrenderer --cflags`
```
Use `$(shell pkg-config ...)` instead.

jserv commented on Makefile:
```diff
@@ -41,6 +44,53 @@ ifeq ($(call has, VIRTIONET), 1)
+endif
+
+ifeq ($(ENABLE_SDL),1)
+ifeq ($(ENABLE_VIRTIOGPU),1)
```
Why did you check `ENABLE_SDL` here?

jserv commented on configs/linux.config:
```diff
@@ -92,18 +96,23 @@ CONFIG_TICK_CPU_ACCOUNTING=y
 # CONFIG_PSI is not set
 # end of CPU/Task time and stats accounting
 
+CONFIG_CPU_ISOLATION=y
```
Change the configurations which are required. Minimize as possible as you can.

jserv commented on scripts/build-image.sh:
```diff
@@ -38,16 +38,48 @@ function do_buildroot
+    popd
+}
+
+function do_directfb_examples
```
Merge the function into `do_directfb`.

shengwen-tw commented:
> `semu` can now run `DirectFB2` with DRM/KMS mode.
> 
> This is done by enforcing DirectFB2 to run with `X8R8G8B8` format via code modification. (By default DirectFB2 uses `A8R8G8B8` format but this failed to work so far, which requires further investigation)
> 
> `DRM/KMS` is crucial for integrating VirGL to the `semu` project as `Linux framebuffer device` does not trigger any 3D rendering operations.

A change to the DirectFB2 is proposed for resolving the issue:
https://github.com/directfb2/DirectFB2/pull/154

jserv reviewed (changes requested):
Rebase the latest `master` branch to pass CI pipeline for macOS.

shengwen-tw commented:
Cross-referencing #68 regarding the unexpected hanging of kmscube.

shengwen-tw commented:
First successful attempt to run `glxgears` using Mesa3D / GLX in a VirtIO-GPU 2D rendering environment.
![image](https://github.com/user-attachments/assets/a8fb9ade-93c4-4ab6-9fb4-485b52b95ccf)

shengwen-tw committed:
Implement virtio-input devices

shengwen-tw committed:
Implement virglrenderer backend for virtio-gpu

Follow the installation guide from:
https://gitlab.freedesktop.org/virgl/virglrenderer

Prerequisite:
`sudo apt install libepoxy-dev`

jserv reviewed (changes requested):
Rebase the latest `master` branch.

shengwen-tw commented:
@jserv :
I propose we begin the code review to merge the current progress in this pull request.
I will open a separate pull request to add support for the 3D rendering mode.

shengwen-tw committed:
linux.config: Enable VirtIO GPU and input support

jserv commented on Makefile:
```diff
@@ -115,6 +116,62 @@ LDFLAGS += -lm
+
+# virtio-gpu and virgl
+ENABLE_VIRTIOGPU ?= 1
+ENABLE_VIRGL ?= 1
```
It is confusing to have VirGL here. Add comments.

shengwen-tw committed:
buildroot.config: Enable libdrm

shengwen-tw committed:
Support building X11 with Buildroot optionally
